common/lib/client/channel.ts

```diff
@@ -0,0 +1,178 @@
+import * as Utils from '../util/utils';
+import EventEmitter from '../util/eventemitter';
+import Logger from '../util/logger';
+import Presence from './presence';
+import Crypto from 'platform-crypto';
+import Message, { CipherOptions } from '../types/message';
+import ErrorInfo from '../types/errorinfo';
+import PaginatedResource, { PaginatedResult } from './paginatedresource';
+import Http from 'platform-http';
+import Resource from './resource';
+import { ChannelOptions } from '../../types/channel';
+import { PaginatedResultCallback } from '../../types/utils';
+
+// TODO: Replace these when Realtime and Rest are in TypeScript
+type Realtime = any;
+type Rest = any;
+
+interface RestHistoryParams {
+	start?: number;
+	end?: number;
+	direction?: string;
+	limit?: number;
+}
+
+function noop() {}
+const MSG_ID_ENTROPY_BYTES = 9;
+
+function allEmptyIds(messages: Array<Message>) {
+	return Utils.arrEvery(messages, function(message: Message) {
+		return !message.id;
+	});
+}
+
+function normaliseChannelOptions(options?: ChannelOptions) {
+	const channelOptions = options || {};
+	if(channelOptions.cipher) {
+		if(!Crypto) throw new Error('Encryption not enabled; use ably.encryption.js instead');
+		const cipher = Crypto.getCipher(channelOptions.cipher);
+		channelOptions.cipher = cipher.cipherParams;
+		channelOptions.channelCipher = cipher.cipher;
+	} else if('cipher' in channelOptions) {
+		/* Don't deactivate an existing cipher unless options
+			* has a 'cipher' key that's falsey */
+		channelOptions.cipher = null;
+		channelOptions.channelCipher = null;
+	}
+	return channelOptions;
+}
+
+class Channel extends EventEmitter {
+	rest: Rest | Realtime;
+	name: string;
+	basePath: string;
+	presence: Presence;
+	channelOptions: ChannelOptions;
+
+	constructor(rest: Rest | Realtime, name: string, channelOptions?: ChannelOptions) {
+		super();
+		Logger.logAction(Logger.LOG_MINOR, 'Channel()', 'started; name = ' + name);
+		this.rest = rest;
+		this.name = name;
+		this.basePath = '/channels/' + encodeURIComponent(name);
+		this.presence = new Presence(this);
+		this.channelOptions = {};
+		this.channelOptions = normaliseChannelOptions(channelOptions);
```

Comment:
Will this second assignment of channelOptions not override the first one ?

Reply:
Resolving this conversation on @ikbalkaya's behalf.